ostree_upload_server/importers/flatpak.py

```diff
@@ -0,0 +1,78 @@
+import logging
+
+import gi
+gi.require_version('OSTree', '1.0')
+from gi.repository import GLib, OSTree
+
+from .base import BaseImporter
+from .util import get_metadata_contents
+
+
+OSTREE_COMMIT_GVARIANT_STRING = "(a{sv}aya(say)sstayay)"
+
+OSTREE_STATIC_DELTA_META_ENTRY_FORMAT = "(uayttay)"
+
+OSTREE_STATIC_DELTA_FALLBACK_FORMAT = "(yaytt)"
+
+OSTREE_STATIC_DELTA_SUPERBLOCK_FORMAT = \
+    "(a{sv}tayay" + OSTREE_COMMIT_GVARIANT_STRING + \
+    "aya" + OSTREE_STATIC_DELTA_META_ENTRY_FORMAT + \
+    "a" + OSTREE_STATIC_DELTA_FALLBACK_FORMAT + \
+    ")"
+
+
+class FlatpakImporter(BaseImporter):
+    MIME_TYPE = 'application/octet-stream'
+
+    METADATA_KEYS = ['ref',
+                     'flatpak',
+                     'origin',
+                     'runtime-repo',
+                     'metadata',
+                     'gpg-keys']
+
+    def _import_commit(self, target_repo, src_path_obj, refs=None):
+        # Apply the delta to the target repo
+        target_repo.static_delta_execute_offline(src_path_obj, False, None)
+
+        # Compare installed and header metadata, remove commit if mismatch (abort transaction)
+        metadata_contents = get_metadata_contents(target_repo, self._commit)
+        if metadata_contents != self._metadata['metadata']:
+            raise Exception("Committed metadata does not match the static delta header")
+
+        logging.debug("Committed metadata matches the static delta header")
+
+    def import_to_repo(self):
+        logging.info("Trying to use %s extractor...", self.__class__.__name__)
+
+        ### Mmap the flatpak file and create a GLib.Variant from it
+        # TODO: MMapping a huge file is a bad idea and we should fix this
```

Comment:
Did you add this TODO? There's nothing wrong with mmaping a large file that I'm aware of. It takes a large amount of virtual memory, but who cares? This is the exact same thing flatpak itself does when loading a bundle.

Reply:
Yeah I added the TODO. Use of RAM arbitrarily makes running things in containers and memory-restricted environments extremely hard and is the exact reason why we need this hack to begin with (ostree mmapping content a bunch of times). As our apps get really large, this code will suffer the same issues and since swap space is not consistently allocated depending on the machine this is running on it made sense to point it out for whoever works on improvements to the code later.

Reply:
But in this case you're actually saving considerable amounts of physical memory by `mmap`ing it. You need to get the entire `GVariant` (the bundle) into memory to parse it. If you actually just read the entire thing into a buffer, you'd need the equivalent in physical memory. Using `mmap` allows it to only load the pages into physical memory on demand. As long as you're then careful about how you manage the `GVariant` (I'm not sure we are), then you can get away with only loading the initial parts into memory.

Reply:
@dbnicholson I would think that you can parse the data chunk-at-a-time until you get the superblock header from it - it would be a real shame if you can't do that :(

I'll remove the note but needing to mmap the whole GVariant just to read its header seems like something we should either optimize or try to avoid using at all.